### ISSUE TITLE
recent changes breaks restart capabilities

### ISSUE THREAD

**Issue by zingale** [AMReX-Codes/amrex#104]
e.g., in Castro, compiling with AMReX `96601a9` checkpointing and restart work fine.  With the current development head for AMReX (`27e6b9a6`), we get a segfaul upon restart.

Test suite example here:

http://bender.astro.sunysb.edu/Castro/test-suite/test-suite-amrex-gfortran/2017-08-26/rad-2Tshock-1d-restart.html

backtrace:

```
=== If no file names and line numbers are shown below, one can run
            addr2line -Cfie my_exefile my_line_address
    to convert `my_line_address` (e.g., 0x4a6b) into file name and line number.

=== Please note that the line number reported by addr2line may not be accurate.
    One can use
            readelf -wl my_exefile | grep my_line_address'
    to find out the offset for that line.

 0: ./Castro1d.gnu.DEBUG.MPI.ex() [0x59f79c]
    amrex::BLBackTrace::print_backtrace_info(_IO_FILE*)
    /home/testing/castro-amrex-tests/AMReX//Src/Base/AMReX_BLBackTrace.cpp:88

 1: ./Castro1d.gnu.DEBUG.MPI.ex() [0x59f689]
    amrex::BLBackTrace::handler(int)
    /home/testing/castro-amrex-tests/AMReX//Src/Base/AMReX_BLBackTrace.cpp:51

 2: /lib64/libc.so.6(+0x35950) [0x7f315d170950]
    ??
    ??:0

 3: ./Castro1d.gnu.DEBUG.MPI.ex() [0x56ab46]
    amrex::FabArray<amrex::FArrayBox>::FabArray(amrex::BoxArray const&, amrex::DistributionMapping const&, int, int, amrex::MFInfo const&, amrex::FabFactory<amrex::FArrayBox> const&)
    /home/testing/castro-amrex-tests/AMReX//Src/Base/AMReX_FabArray.H:881

 4: ./Castro1d.gnu.DEBUG.MPI.ex() [0x56514b]
    amrex::MultiFab::MultiFab(amrex::BoxArray const&, amrex::DistributionMapping const&, int, int, amrex::MFInfo const&, amrex::FabFactory<amrex::FArrayBox> const&)
    /home/testing/castro-amrex-tests/AMReX//Src/Base/AMReX_MultiFab.cpp:380

 5: ./Castro1d.gnu.DEBUG.MPI.ex() [0x5f7182]
    amrex::StateData::restartDoit(std::istream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    /home/testing/castro-amrex-tests/AMReX//Src/Amr/AMReX_StateData.cpp:239

 6: ./Castro1d.gnu.DEBUG.MPI.ex() [0x5f6f4f]
    amrex::StateData::restart(std::istream&, amrex::Box const&, amrex::BoxArray const&, amrex::DistributionMapping const&, amrex::StateDescriptor const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    /home/testing/castro-amrex-tests/AMReX//Src/Amr/AMReX_StateData.cpp:220

 7: ./Castro1d.gnu.DEBUG.MPI.ex() [0x5e1497]
    amrex::AmrLevel::restart(amrex::Amr&, std::istream&, bool)
    /home/testing/castro-amrex-tests/AMReX//Src/Amr/AMReX_AmrLevel.cpp:332 (discriminator 2)

 8: ./Castro1d.gnu.DEBUG.MPI.ex() [0x434a4b]
    Castro::restart(amrex::Amr&, std::istream&, bool)
    /home/testing/castro-amrex-tests/Castro//Source/driver/Castro_io.cpp:98

 9: ./Castro1d.gnu.DEBUG.MPI.ex() [0x5c797b]
    amrex::Amr::restart(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    /home/testing/castro-amrex-tests/AMReX//Src/Amr/AMReX_Amr.cpp:1450

10: ./Castro1d.gnu.DEBUG.MPI.ex() [0x5c5d5e]
    amrex::Amr::init(double, double)
    /home/testing/castro-amrex-tests/AMReX//Src/Amr/AMReX_Amr.cpp:1045

11: ./Castro1d.gnu.DEBUG.MPI.ex() [0x43bbf6]
    main
    /home/testing/castro-amrex-tests/Castro//Source/driver/main.cpp:120

12: /lib64/libc.so.6(__libc_start_main+0xf1) [0x7f315d15b401]
    ??
    ??:0

13: ./Castro1d.gnu.DEBUG.MPI.ex() [0x40acea]
    _start
    ??:?
```

**Comment by zingale** [comment]
I put this here, since I think this is an AMReX issue and not a Castro issue

**Comment by zingale** [comment]
a `git bisect` shows that this is the culprit commit:

```
commit 459e4b2ce1c50b0e15986554a8f723cac2dc2bbb
Author: Weiqun Zhang <weiqunzhang@lbl.gov>
Date:   Fri Jul 28 23:49:31 2017 -0700

    add factory and eb to AmrLevel and StateData

:040000 040000 044441b723778e9e5cfc62372c6c7eaaeb45c3b1 94fb2a404f832d2f2104e4627e5d045caa27ed5e M	Src
```

**Comment by WeiqunZhang** [comment]
I think it's fixed now by fe1c54d09c98d.

**Comment by zingale** [comment]
yep, this is fixed now.  Thanks.
